design/nfs/nfs-controller-runtime.md

```diff
@@ -0,0 +1,161 @@
+
+### Validation
+
+CustomResource validation in the operator can be done through the Controller itself. However, the operator pattern has two common types to validate the CustomResource.
+
+- **Syntactic validation** By defining OpenAPI validation rules.
+- **Semantic Validation** By creating a ValidatingAdmissionConfiguration and Admission Webhook.
+
+The current implementation only validates the CustomResource in the controller and just gives an error log in the operator stdout if the given resource is invalid. In this implementation will also cover the CustomResouce validation both though `Syntactic validation` and `Semantic Validation` and also give an improvement validation in the controller.
+
+To implement `Syntactic validation` is only by defining OpenAPI validation rules. Otherwise, the `Semantic Validation` implementation is a bit more complicated. Fortunately, controller-runtime provides an awesome package to help to implement Admission Webhook. Since the webhook server must be served through TLS, a valid TLS certificate will be required. In this case, we can depend on [cert-manager](https://cert-manager.io/)
+
+Example self signed certificate.
+
+```yaml
+---
+apiVersion: cert-manager.io/v1alpha2
+kind: Certificate
+metadata:
+  name: rook-nfs-webhook-cert
+  namespace: rook-nfs-system
+spec:
+  dnsNames:
+  - rook-nfs-webhook.rook-nfs-system.svc
+  - rook-nfs-webhook.rook-nfs-system.svc.cluster.local
+  issuerRef:
+    kind: Issuer
+    name: rook-nfs-selfsigned-issuer
+  secretName: rook-nfs-webhook-cert
+---
+apiVersion: cert-manager.io/v1alpha2
+kind: Issuer
+metadata:
+  name: rook-nfs-selfsigned-issuer
+  namespace: rook-nfs-system
+spec:
+  selfSigned: {}
+```
+
+And the ValidatingAdmissionConfiguration will look like
+
+```yaml
+---
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
+metadata:
+  annotations:
+    cert-manager.io/inject-ca-from: rook-nfs-system/rook-nfs-webhook-cert
+  creationTimestamp: null
+  name: rook-nfs-validating-webhook-configuration
+webhooks:
+- clientConfig:
+    caBundle: Cg==
+    service:
+      name: rook-nfs-webhook
+      namespace: rook-nfs-system
+      path: /validate-nfs-rook-io-v1alpha1-nfsserver
+  failurePolicy: Fail
+  name: validation.nfsserver.nfs.rook.io
+  rules:
+  - apiGroups:
+    - nfs.rook.io
+    apiVersions:
+    - v1alpha1
+    operations:
+    - CREATE
+    - UPDATE
+    resources:
+    - nfsservers
+```
+
+By providing `cert-manager.io/inject-ca-from` annotation, `cert-manager` will replace `.clientConfig.caBundle` with appropriate certificate. And the implementation is the admission webhook server will be deployed independently. So the `Semantic Validation` will be optional and users can enable or disable this validation by deploying the admission webhook server or not. The example manifests to deploy the admission webhook server will look like this.
+
+```yaml
```

Comment:
What will the new installation experience look like if they want to include validation (and all its webhook/cert stuff) too? i.e. what will the new commands to deploy NFS look like, compared to the current installation experience at https://rook.io/docs/rook/v1.3/nfs.html?